### PR TITLE
Using implicit serialization from ModelSerializer

### PR DESCRIPTION
**Pull request by bryan-brancotte**
Hi Jon

I took a look at the serializers, many fields you wrote are implicite as you used a ModelSerializer (ModelSerializer rocks!). 

For example there is a field for shortName [here](https://github.com/joncison/ifbcat-sandbox/blob/master/ifbcatsandbox_api/serializers.py#L219) but as it is already in the [meta](https://github.com/joncison/ifbcat-sandbox/blob/master/ifbcatsandbox_api/serializers.py#L275) it is redundant as the rest-framework automatically generate it from the specification of your model. In the case of shortName it is even error prone as the serializer specify `allow_blank=False` while it is `allow_blank=True` in the [model](https://github.com/joncison/ifbcat-sandbox/blob/master/ifbcatsandbox_api/models.py#L258). 

For EventKeywordSerializer the UniqueValidator is automatically generated from the model, so no need to write it manually.

Here are the fields that can be removed.

In short with ModelSerializer only the CreatableSlugRelatedField should be kept ! (they are very useful, and improve readability of the api responses :+1: )

### CODE REVIEW

**Comment by joncison** [comment]
Thanks @bryan-brancotte, this is great, a huge improvement.  I was wondering about all the (error-prone) duplication with models.py.

But one issue.  The ``Type``, ``Cost`` and ``Accessibility`` fields were serialising with their value (e.g. ``FA``) rather than human-readable name (e.g. ``Free to academics``).  So I reinstated (uncommented) the the code in ``EventSerializer`` for those fields.

Is there a more elegant solution?

PS.  the duplication of the controlled vocabularies (for event type, cost and accessibility) in serializers.py and models.py is bad, and will need to be refactored out, one way or another.

**Comment by bryan-brancotte** [comment]
I checked out the latest version and still see the code for the field like Cost. With or without the serializers.ChoiceField I get the code. To be sure that I did not introduce a regression I checkout the code of [july14th](https://github.com/joncison/ifbcat-sandbox/commit/f42e30e64fb4e52b2c7f9accfb41075bd1f941fc) and still get `"accessibility": "PR",`.

Maybe you are talking of the use case where you PUT/PATCH an event rather than GET it ?

**Comment by joncison** [comment]
PUT/PATCH and GET all yield the human-readable names.   

What happens when (in the browsable API) you reset a value *e.g." from ``PR`` to "Private" - assuming you get an option to select the human-readable names) ?

**Comment by bryan-brancotte** [comment]
Ok I got it:

 * In models.py, event takes values from EventType, `Workshop` is the human readable value, but the value stored in the DB **according to models.py** is `WO`
 * In serializer.py event takes values from `('Workshop', 'Training course', 'Meeting', 'Conference')` thus no mention of `WO`, so when saving from the API (i.e: http://127.0.0.1:8092/api/event/1/) `'Workshop'` is saved in the db
 * In models.py, event have `max_length=2,` but as we use sqlite3, the maximum length is ignored (credit to Hervé who spoted this problem on an other project)
 * Thus when saving from the API, you save in the DB `'Workshop'`, causing the problem you have in https://github.com/joncison/ifbcat-sandbox/pull/8#issuecomment-659337454 as django search for entry with `WO` as type in the DB

The next step is to find a way to use the human readable value in the API while storing the code in the DB, an issue we are not the only one to have (https://stackoverflow.com/questions/60940861/accessing-django-3-textchoices-via-an-exposed-api-endpoint)

**Comment by joncison** [comment]
Ahh, I see.

We also need to be mindful that the shortname e.g. ``FA`` is not always the first two characters of the human-readable label e.g. "Free to academics".

**Comment by joncison** [comment]
@bryan-brancotte  - would a simple solution to be to replace the shortnames with the human-readable labels (which then would not be needed), e.g.

```
class EventType(models.TextChoices):
   WORKSHOP = 'Workshop'
   TRAINING_COURSE = 'Training course')
   .. 
```

?

I can't imagine (but could be wrong) this would cause a performance hit, or other problem?
But I'm assuming the short names can include space characters ...

**Comment by bryan-brancotte** [comment]
Replace the shortnames with the human-readable labels could be a solution, but the field would have to be enlarge, and when filtering it would be over string, not very small string.

There are solutions (https://github.com/encode/django-rest-framework/issues/1755#issuecomment-193018102, https://stackoverflow.com/questions/28945327/django-rest-framework-with-choicefield) but I am not satified of them.

On the other hand, the OPTION verb allows to get all information on type:
```
 "type": {
                "type": "choice",
                "required": false,
                "read_only": false,
                "label": "Type",
                "choices": [
                    {
                        "value": "WO",
                        "display_name": "Workshop"
                    },
                    {
                        "value": "TR",
                        "display_name": "Training course"
                    },
                    ...
                ]
            },
```
So the question is to know if user in the end will manually use the UI served at /api/. If only program access it, these application will have to parse OPTION results to know all the values for type, and use the display_name to render, and the value to PUT. I would even recommend to use an models.IntegerChoices to only manipulate int in the db.

Also we can add in the serializer
```
type_str = serializers.CharField(source='get_type_display',read_only=True)
```
which will show the display_name for type, while also exposing the code, and letting user PUT with the value (WO):
```
HTTP 200 OK
Allow: GET, PUT, PATCH, DELETE, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "id": 1,
    "user_profile": 1,
    "name": "event name",
    "shortName": "",
    "description": "blabla",
    "homepage": "",
    "type": "WO",
    "type_str": "Workshop",
    "venue": "",
    "city": "",
    "country": "",
    "onlineOnly": true,
```

**Comment by joncison** [comment]
Thanks Bryan for looking into this.

Users may end up using the /api/ UI, at least until a better front-end is available.  They might also (eventually) be working with JSON files directly in GitHub.  So I need the JSON to be fully human-readable.

So I'll experiment with one solution or another .... strikes me the simplest thing might be just to use the long strings only as suggested (unless you know already why this would be bad?)

**Comment by joncison** [comment]
I tested above out and it works OK.  Please let me know if this is "good enough" (for now at least).  Thanks!

**Comment by bryan-brancotte** [comment]
I works and fulfill the need of an API fully human-readable, so I think we can say that it is good enough.

The point I would raise is that if we set up an index on one of this field, it is an index over string of 255 char, and not over 2 char or 1 int, so bigger, and maybe slower. similarily when filterig it is a string comparison, not just an int comparison. The question is really the amount of information stored in this application, will it be hundreds thousands of event or just thousand(s)? I think we can consider it ok, and if you experience performance issue, then rework this aspect.

Just to keep note of idea, an alternative solution would be to replace EventType by a Model and use a SlugRelatedField in the serializer to show and get them. I don't know how/if filtering would work with this solution.
